### ISSUE TITLE
Ran out of repos.... am I doing this wrong?

### ISSUE THREAD

**Issue by mafudge** [education/teachers#28]
I have a sandboxed repository setup of 40 students. I'm giving out 7 assignments. 

Right now my plan is 1 repo per student / assignment. = 280 repos. I was only given a 50 repository plan. Is there a better way to do this? 

I don't have all the assignments figured out yet. I assume I cannot simply use 1 repository for all 7 assignments, correct? 

I cannot push_files after students have started cloning the repository?

**Comment by efthym** [comment]
I recall reading somewhere,  a while ago,that you could request for more and that the request is usually granted.

If you can't get more:
- ask the students to clone their work, so they can keep a copy
- clone them yourself (teachers_pet clone_repos) for marking locally
- delete the old repos. I don't think teachers_pet can do this though.
- create new ones.

Any other ideas are welcome, as I expect to run into similar problems soon.

**Comment by tarebyte** [comment]
@mafudge, if your needs are larger than what you currently have I would email education@github.com and they should be able to help you out.

To answer your repo question. 

If you push a second time after students have worked on their repo you would have to force push it. Which `push_files` does not do.

From the code (wasn't sure if you'd be interested in this or not): 
```ruby
puts "Adding remotes and pushing files to student repositories."
remotes_to_add.keys.each do |remote|
  puts "#{remote} --> #{remotes_to_add[remote]}"
  `git remote add #{remote} #{remotes_to_add[remote]}`
  `git push #{remote} master`
end
```

Can be located here:
https://github.com/education/teachers_pet/blob/fa6b00a20d54b4ad389f9836fa78645736932d9c/lib/teachers_pet/actions/push_files.rb

**Comment by aleksandar-todorovic** [comment]
I don't think you need 280 repositories to do so. My suggestion is 40 different repositories (one for each student), then create a specific folder inside of their repository for each assignment (so seven folders in each repository).

Let the students know that they should _not_ mess with the folder structure and share the specific repository with the specific student. Of course, only that student needs to have a write permission to this repository and they should be private.

After doing so, you can update their repository to add a new assignment and they can push their assignment once they are done with them.

I think that this solution is much easier than navigating through 280 different repositories. I hope you'll agree with me.

**Comment by efthym** [comment]
I finished a large course with 200 students x 8 assignments about a month ago and I can share how I handled this situation.

I planned to follow what I proposed in my previous post:
- I asked github for 2 x Number of students private repositories (and they kindly did it)
- I would then be cloning and marking the 1st assignment while the students work on the 2nd.
  By the time I had finished (and before the deadline of the 2nd assignment), I'd delete the 1st assignments repos and create new ones for the 3rd.

However I run into [Berkeley Uni's CS51C course] (http://www-inst.eecs.berkeley.edu/~cs61c) and followed their idea (which is also what aleksandar proposed in the previous post):
- one private repository per student
- one **public repository** for the "starter files" of each assignment
- use one directory/folder per assignment under the private repositories

So for each assignment, I prepared a public starter repository and ask students to merge it into their repositories to get started.
Students had to do something like
```
git remote add lab01_starter https://github.com/ORG_NAME/lab01_starter.git
git fetch lab01_starter
git merge lab01_starter/master -m "Fetched lab01 starter files"
```
This created a lab01 directory in their repo where they found the skeleton and the detailed assignment description.This is also where they had to place their answers before the assignment deadline.

I had to ask the students to follow this directory structure and not create similarly named directories.
A few did use different names in the first assignments and I had to email/open issues to get them to follow the norm.

I also modified the teachers_pet clone_repos command to actually do a pull instead of a clone after the first time I got student's assignments, so as to speed up download time to my machine for marking.

**Comment by sjm324** [comment]
@efthym Thank you for sharing this!  My course has ~200 students, and either 4 or 5 assignments.  Asking for 800 - 1000 repos hardly seems appropriate; your approach confirms that I can just make one repo for all of their assignments :)  Thanks!
